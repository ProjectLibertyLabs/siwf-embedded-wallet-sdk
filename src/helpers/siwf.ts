import { SignatureFn } from "../types";
import { SiwfResponsePayload, SiwfResponse } from "@projectlibertylabs/siwf";
import { encodedValueToSignature, accountIdToPublicKey } from "./utils";

export interface CreateLoginSiwfResponseArguments
  extends Record<string, unknown> {
  domain: string;
  uri: string;
  version: string;
  chainId: string;
  nonce?: string;
  issuedAt?: string;
}

export async function createLoginSiwfResponse(
  accountId: string,
  signatureFn: SignatureFn,
  payloadArguments: CreateLoginSiwfResponseArguments,
): Promise<SiwfResponse> {
  const loginCaip122 = `${payloadArguments.domain} wants you to sign in with your Frequency account:
    frequency:${payloadArguments.chainId}:${accountId}
    
    URI: ${payloadArguments.uri}
    Version: ${payloadArguments.version}
    Nonce: ${payloadArguments.nonce}
    Chain ID: frequency:${payloadArguments.chainId}
    Issued At: ${payloadArguments.issuedAt}`;

  const encodedValue = await signatureFn({
    method: "personal_sign",
    params: [accountId, loginCaip122],
  });

  const userPublicKey = accountIdToPublicKey(accountId);

  const signature = encodedValueToSignature(encodedValue);

  return {
    userPublicKey,
    payloads: [
      {
        signature,
        type: "login",
        payload: {
          message: loginCaip122,
        },
      },
    ],
    credentials: [],
  };
}

export async function createSignInSiwfResponse(
  accountId: string,
  payloads: SiwfResponsePayload[],
): Promise<SiwfResponse> {
  const userPublicKey = accountIdToPublicKey(accountId);
  return {
    userPublicKey,
    payloads,
    credentials: [],
  };
}
