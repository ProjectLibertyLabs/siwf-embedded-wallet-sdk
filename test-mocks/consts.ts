import {
  AccountResponse,
  ChainInfoResponse,
  GatewaySiwfResponse,
} from "../src/gateway-types";
import { mockGatewayNewUserResponse } from "./gateway-new-user";

// Matching mock user addresses
export const mockUserAddress = "0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac";
export const mockControlKey =
  "5HYRCKHYJN9z5xUtfFkyMj4JUhsAwWyvuU8vKB1FcnYTf9ZQ";

// Provider Mocks
export const mockProviderControlKey =
  "f6cL4wq1HUNx11TcvdABNf9UNXXoyH47mVUwT59tzSFRW8yDH";
export const mockProviderEncodedRequest =
  "eyJyZXF1ZXN0ZWRTaWduYXR1cmVzIjp7InB1YmxpY0tleSI6eyJlbmNvZGVkVmFsdWUiOiJmNmNMNHdxMUhVTngxMVRjdmRBQk5mOVVOWFhveUg0N21WVXdUNTl0elNGUlc4eURIIiwiZW5jb2RpbmciOiJiYXNlNTgiLCJmb3JtYXQiOiJzczU4IiwidHlwZSI6IlNyMjU1MTkifSwic2lnbmF0dXJlIjp7ImFsZ28iOiJTcjI1NTE5IiwiZW5jb2RpbmciOiJiYXNlMTYiLCJlbmNvZGVkVmFsdWUiOiIweDA0MDdjZTgxNGI3Nzg2MWRmOTRkMTZiM2ZjYjMxN2QzN2EwN2FiYzJhN2Y5Y2Q3YzAyY2MyMjUyOWVlN2IzMmQ1Njc5NWY4OGJkNmI0YWQxMDZiNzJiOTFiNjI0NmE3ODM2NzFiY2QyNGNiMDFhYWYwZTkzMTZkYjVlMGNkMDg1In0sInBheWxvYWQiOnsiY2FsbGJhY2siOiJodHRwOi8vbG9jYWxob3N0OjMwMDAiLCJwZXJtaXNzaW9ucyI6WzUsNyw4LDksMTBdfX0sInJlcXVlc3RlZENyZWRlbnRpYWxzIjpbeyJ0eXBlIjoiVmVyaWZpZWRHcmFwaEtleUNyZWRlbnRpYWwiLCJoYXNoIjpbImJjaXFtZHZteGQ1NHp2ZTVraWZ5Y2dzZHRvYWhzNWVjZjRoYWwydHMzZWV4a2dvY3ljNW9jYTJ5Il19LHsiYW55T2YiOlt7InR5cGUiOiJWZXJpZmllZEVtYWlsQWRkcmVzc0NyZWRlbnRpYWwiLCJoYXNoIjpbImJjaXFlNHFvY3poZnRpY2k0ZHpmdmZiZWw3Zm80aDRzcjVncmNvM29vdnd5azZ5NHluZjQ0dHNpIl19LHsidHlwZSI6IlZlcmlmaWVkUGhvbmVOdW1iZXJDcmVkZW50aWFsIiwiaGFzaCI6WyJiY2lxanNwbmJ3cGMzd2p4NGZld2NlazVkYXlzZGpwYmY1eGppbXo1d251NXVqN2UzdnUydXducSJdfV19XX0";

export const mockProviderEncodedRequestWithoutGraphKey =
  "eyJyZXF1ZXN0ZWRTaWduYXR1cmVzIjp7InB1YmxpY0tleSI6eyJlbmNvZGVkVmFsdWUiOiJmNmNMNHdxMUhVTngxMVRjdmRBQk5mOVVOWFhveUg0N21WVXdUNTl0elNGUlc4eURIIiwiZW5jb2RpbmciOiJiYXNlNTgiLCJmb3JtYXQiOiJzczU4IiwidHlwZSI6IlNyMjU1MTkifSwic2lnbmF0dXJlIjp7ImFsZ28iOiJTcjI1NTE5IiwiZW5jb2RpbmciOiJiYXNlMTYiLCJlbmNvZGVkVmFsdWUiOiIweDA0MDdjZTgxNGI3Nzg2MWRmOTRkMTZiM2ZjYjMxN2QzN2EwN2FiYzJhN2Y5Y2Q3YzAyY2MyMjUyOWVlN2IzMmQ1Njc5NWY4OGJkNmI0YWQxMDZiNzJiOTFiNjI0NmE3ODM2NzFiY2QyNGNiMDFhYWYwZTkzMTZkYjVlMGNkMDg1In0sInBheWxvYWQiOnsiY2FsbGJhY2siOiJodHRwOi8vbG9jYWxob3N0OjMwMDAiLCJwZXJtaXNzaW9ucyI6WzUsNyw4LDksMTBdfX0sInJlcXVlc3RlZENyZWRlbnRpYWxzIjpbeyJhbnlPZiI6W3sidHlwZSI6IlZlcmlmaWVkRW1haWxBZGRyZXNzQ3JlZGVudGlhbCIsImhhc2giOlsiYmNpcWU0cW9jemhmdGljaTRkemZ2ZmJlbDdmbzRoNHNyNWdyY28zb292d3lrNnk0eW5mNDR0c2kiXX0seyJ0eXBlIjoiVmVyaWZpZWRQaG9uZU51bWJlckNyZWRlbnRpYWwiLCJoYXNoIjpbImJjaXFqc3BuYndwYzN3ang0ZmV3Y2VrNWRheXNkanBiZjV4amltejV3bnU1dWo3ZTN2dTJ1d25xIl19XX1dfQ"

// Returning User Mocks
export const mockProviderAccountResponse: AccountResponse = {
  msaId: "1",
};

export const mockReturningUserAccountResponse: AccountResponse = {
  msaId: "290",
  handle: {
    base_handle: "Bob",
    canonical_base: "Bob",
    suffix: 0,
  },
};

export const mockRetunringUserGatewaySiwfResponse: GatewaySiwfResponse = {
  controlKey: mockControlKey,
  msaId: mockReturningUserAccountResponse.msaId,
};

// New User Mocks
export const mockNewUserAccountResponse = null;

export const mockNewUserGatewaySiwfResponse: GatewaySiwfResponse = {
  ...mockGatewayNewUserResponse(),
  controlKey: mockControlKey,
};

export const mockChainInfoResponse: ChainInfoResponse = {
  blocknumber: 32,
  finalized_blocknumber: 32,
  genesis: "0x4a587bf17a404e3572747add7aab7bbe56e805a5479c6c436f07f36fcc8d3ae1", // From mainnet
  runtime_version: 2,
};
