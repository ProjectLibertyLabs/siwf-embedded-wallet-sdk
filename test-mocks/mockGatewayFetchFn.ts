import { GatewayFetchFn } from "../src/types/param-types";
import {
  AccountResponse,
  ChainInfoResponse,
  GatewaySiwfResponse,
} from "../src/types/response-types";

export const mockGatewayFetchFactory = (
  hasAccountResponse: AccountResponse | null,
  providerAccountResponse: AccountResponse | null,
  gatewaySiwfResponse: GatewaySiwfResponse,
  chainInfoResponse: ChainInfoResponse,
  providerControlKey: string,
  finalResponse: { response: AccountResponse | null } = { response: null },
): GatewayFetchFn => {
  return async (method, path) => {
    if (path === "/v2/accounts/siwf") {
      return {
        ok: true,
        status: 200,
        json: async () => gatewaySiwfResponse,
      } as Response;
    }

    if (path.startsWith("/v1/accounts/account/")) {
      if (method === "GET")
        return {
          ok: true,
          status: 200,
          json: async () => {
            if (path.includes(providerControlKey))
              return providerAccountResponse;
            return finalResponse.response || hasAccountResponse;
          },
        } as Response;
    }

    if (path.startsWith("/v1/frequency/blockinfo")) {
      if (method === "GET")
        return {
          ok: true,
          status: 200,
          json: async () => chainInfoResponse,
        } as Response;
    }

    return new Response(`Error: ${path} not found`, {
      status: 404,
    });
  };
};
