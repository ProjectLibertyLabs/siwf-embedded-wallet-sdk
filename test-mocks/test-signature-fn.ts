import { TypedDataField, Wallet } from "ethers";
import { SignatureFn } from "../src/types/param-types";

/**
 * Test signature function
 * Signs using 0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac
 * @param request
 * @returns string
 */
export const TEST_SIGNATURE_FN: SignatureFn = async (request) => {
  // Alith: 0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac
  const privateKey =
    "0x5fb92d6e98884f76de468fa3f6278f8807c48bebc13595d45af5bdc4da702133";
  const wallet = new Wallet(privateKey);

  if (request.method === "personal_sign") {
    const [_address, message] = request.params;
    return await wallet.signMessage(message);
  }
  if (request.method === "eth_signTypedData_v4") {
    const [_address, typedData] = request.params;

    // Ethers forces this type to be removed
    const types: Record<string, TypedDataField[]> = structuredClone(
      typedData.types,
    );
    delete types["EIP712Domain"];

    return await wallet.signTypedData(
      typedData.domain,
      types,
      typedData.message,
    );
  }
  throw new Error("Incorrect method sent to the TEST_SIGNATURE_FN");
};
