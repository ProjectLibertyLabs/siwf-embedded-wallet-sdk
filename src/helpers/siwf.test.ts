import { describe, expect, it } from "vitest";
import { TEST_SIGNATURE_FN } from "../../test-mocks/test-signature-fn";
import {
  createLoginSiwfResponse,
  CreateLoginSiwfResponseArguments,
} from "./siwf";
import { SiwfResponse } from "@projectlibertylabs/siwf";

describe("createSignedLogInPayload", () => {
  it("returns the correct payload", async () => {
    const accountId = "0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac";
    const signatureFn = TEST_SIGNATURE_FN;
    const mockLoginPayloadArguments: CreateLoginSiwfResponseArguments = {
      domain: "your-app.com",
      uri: "https://your-app.com/signin/callback",
      version: "1",
      nonce: "N6rLwqyz34oUxJEXJ",
      chainId: "123",
      issuedAt: "2024-10-29T19:17:27.077Z",
    };

    const payload: SiwfResponse = await createLoginSiwfResponse(
      accountId,
      signatureFn,
      mockLoginPayloadArguments,
    );

    expect(payload).toMatchSnapshot();
  });
});
