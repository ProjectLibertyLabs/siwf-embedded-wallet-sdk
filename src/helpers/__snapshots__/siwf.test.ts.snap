// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSignedLogInPayload > returns the correct payload 1`] = `
{
  "credentials": [],
  "payloads": [
    {
      "payload": {
        "message": "your-app.com wants you to sign in with your Frequency account:
frequency:123:0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac

URI: https://your-app.com/signin/callback
Version: 1
Nonce: N6rLwqyz34oUxJEXJ
Chain ID: frequency:123
Issued At: 2024-10-29T19:17:27.077Z",
      },
      "signature": {
        "algo": "SECP256K1",
        "encodedValue": "0x461e9584dc25ea8fd30ab00b6b1df5d313e8e7bbc1cee24950376aa3e416645137984b2ebd9bbb40ecc2f79b2dbff0eff96d689db6267c216ea2cb074931006e1b",
        "encoding": "base16",
      },
      "type": "login",
    },
  ],
  "userPublicKey": {
    "encodedValue": "0xf24FF3a9CF04c71Dbc94D0b566f7A27B94566cac",
    "encoding": "base16",
    "format": "eip-55",
    "type": "Secp256k1",
  },
}
`;
